editors/code/src/inlay_hints.ts

```diff
@@ -1,162 +1,319 @@
+import * as lc from "vscode-languageclient";
 import * as vscode from 'vscode';
 import * as ra from './rust-analyzer-api';
 
 import { Ctx } from './ctx';
-import { log, sendRequestWithRetry } from './util';
+import { sendRequestWithRetry, assert } from './util';
 
 export function activateInlayHints(ctx: Ctx) {
-    const hintsUpdater = new HintsUpdater(ctx);
+    const hintsUpdater = new HintsUpdater(ctx.client);
+
     vscode.window.onDidChangeVisibleTextEditors(
-        async _ => hintsUpdater.refresh(),
+        visibleEditors => hintsUpdater.refreshVisibleRustEditors(
+            visibleEditors.filter(isRustTextEditor)
+        ),
         null,
         ctx.subscriptions
     );
 
     vscode.workspace.onDidChangeTextDocument(
-        async event => {
-            if (event.contentChanges.length === 0) return;
-            if (event.document.languageId !== 'rust') return;
-            await hintsUpdater.refresh();
+        ({ contentChanges, document }) => {
+            if (contentChanges.length === 0) return;
+            if (!isRustTextDocument(document)) return;
+
+            hintsUpdater.refreshRustDocument(document);
         },
         null,
         ctx.subscriptions
     );
 
     vscode.workspace.onDidChangeConfiguration(
-        async _ => hintsUpdater.setEnabled(ctx.config.displayInlayHints),
+        async _ => {
+            // FIXME: ctx.config may have not been refreshed at this point of time, i.e.
+            // it's on onDidChangeConfiguration() handler may've not executed yet
+            // (order of invokation is unspecified)
+            // To fix this we should expose an event emitter from our `Config` itself.
```

Comment:
I guess a better approach is to just make `ctx.config` **always** reflect the most recent config? Ie, we should rid of `config.onDidChange` handler.

Reply:
Yeah, we will get rid of cache

Reply:
Though, we won't get rid of `onConfigChange`, since it triggerses (`option requires reload`) message and updates logging `setEnabled()`. Switching to event emitter for config is actually very simple, vscode exposes `vscode.EventEmitter` for this.